### PR TITLE
tvd-section: fix creation algorithm

### PR DESCRIPTION
**Pull request by flomonster**

### CODE REVIEW

File: src/main/java/fr/sncf/osrd/infra/routegraph/RouteGraph.java [hunk 1]
```diff
@@ -69,80 +71,49 @@ public Route makeRoute(
...
-                    }
-                }
-                if (copiedTvdSections.size() != tvdSections.size())
+                // Checks the path goes through all TVDSections
+                var pathTvdSections = new SortedArraySet<TVDSection>();
+                for (var tvdSectionPath : tvdSectionsPath)
+                    pathTvdSections.add(tvdSectionPath.tvdSection);
+                if (pathTvdSections.size() != tvdSections.size())
                     throw new InvalidInfraException(
                             "Route has a tvd section that is not part of any of its release groups");
```
Comment (multun) [inline src/main/java/fr/sncf/osrd/infra/routegraph/RouteGraph.java:80]: this doesn't feel right, why would that be true

**Review by multun — APPROVED** [review]

File: src/main/java/fr/sncf/osrd/infra/routegraph/RouteGraph.java (hunk 1 above)
Comment (flomonster) [reply src/main/java/fr/sncf/osrd/infra/routegraph/RouteGraph.java:80]: Currently, we don't support routes that pass through the same tvd section multiple times. This condition ensures that.

File: src/main/java/fr/sncf/osrd/infra/routegraph/RouteGraph.java [hunk 2]
```diff
@@ -40,24 +33,33 @@
      */
     public List<Route> getIncomingNeighbors(Route route) {
         var firstTvdSectionPath = route.tvdSectionsPaths.get(0);
-        var firstTvdSectionPathDir = route.tvdSectionsPathDirections.get(0);
-        var node = getNode(firstTvdSectionPath.getStartNode(firstTvdSectionPathDir));
-        var nodeDirection = firstTvdSectionPath.nodeDirection(firstTvdSectionPathDir, EdgeEndpoint.BEGIN);
-        return node.getIncomingRouteNeighbors(nodeDirection);
+        var firstWaypoint = firstTvdSectionPath.startWaypoint;
+        return firstWaypoint.getIncomingRouteNeighbors(firstTvdSectionPath.getStartTrackDirection());
     }
```
Comment (eckter) [inline src/main/java/fr/sncf/osrd/infra/routegraph/RouteGraph.java]: This method seems to be unused, is there a reason we keep it?

File: src/main/java/fr/sncf/osrd/infra/routegraph/RouteGraph.java (hunk 1 above)
Comment (eckter) [reply src/main/java/fr/sncf/osrd/infra/routegraph/RouteGraph.java:80]: Does the condition really ensure that? It's a set so if we pass through the same section twice, it's only added once.  This only checks that we pass trough each section at least once.

File: src/main/java/fr/sncf/osrd/infra/routegraph/RouteGraph.java [hunk 3]
```diff
@@ -177,128 +147,133 @@ public RouteGraph build() {
...
+                String routeId,
+                TrackSection edge,
+                EdgeDirection direction,
+                Map<Switch, SwitchPosition> switchPositions
+        ) throws InvalidInfraException {
+            var trackSectionEndNode = direction == START_TO_STOP ? edge.endNode : edge.startNode;
+            var neighbors = direction == START_TO_STOP ? edge.endNeighbors : edge.startNeighbors;
+            if (neighbors.size() == 0)
+                throw new InvalidInfraException(
+                        "couldn't reach the endwaypoint of route " + routeId
```
Comment (eckter) [inline src/main/java/fr/sncf/osrd/infra/routegraph/RouteGraph.java]:
```suggestion
                        "couldn't reach the end waypoint of route " + routeId
```

File: src/main/java/fr/sncf/osrd/railml/RMLRoute.java [hunk 4]
```diff
@@ -56,7 +56,11 @@
             if (entrySignal != null)
                 entrySignal.linkedDetector = new ID<>(entryWaypoint.id);
 
-            res.add(new RJSRoute(id, switchesPosition, releaseGroups, entryWaypoint));
+            // TODO: Compute the entry direction using the signal direction or buffer stop position
```
Comment (eckter) [inline src/main/java/fr/sncf/osrd/railml/RMLRoute.java]: There's a TODO left, we should open an issue for that

File: src/main/java/fr/sncf/osrd/infra/routegraph/RouteGraph.java (hunk 1 above)
Comment (multun) [reply src/main/java/fr/sncf/osrd/infra/routegraph/RouteGraph.java:80]: it does, the number of tvdsectionpaths counts how many times we changed tvdsections. if we changed tvdsections more often than there are tvdsections, then we went through the same tvdsection multiple times

File: src/main/java/fr/sncf/osrd/infra/routegraph/RouteGraph.java (hunk 2 above)
Comment (multun) [reply src/main/java/fr/sncf/osrd/infra/routegraph/RouteGraph.java]: nop, it's gone now

File: src/main/java/fr/sncf/osrd/railml/RMLRoute.java (hunk 4 above)
Comment (multun) [reply src/main/java/fr/sncf/osrd/railml/RMLRoute.java]: done

**Pull request merged**